### PR TITLE
Use structures to describe STDP config

### PR DESCRIPTION
**Pull request by dkfellows**
Part of my long-term plan to expunge plain address arithmetic; part of what I was doing in #857

### CODE REVIEW

**Review by andrewgait — APPROVED** [review]
This looks fine to me, though I didn't notice a sPyNNaker8 branch to run the integration tests for it?

**Comment by dkfellows** [comment]
The trigger's now cooking

File: neural_modelling/src/neuron/plasticity/stdp/timing_dependence/timing_vogels_2011_impl.c [hunk 1]
```diff
@@ -28,18 +28,25 @@ int16_lut *tau_lookup;
...
     log_info("\tVogels 2011 timing rule");
+    vogels_2011_config_t *config = (vogels_2011_config_t *) address;
 
     // Copy parameters
-    plasticity_trace_region_data.alpha = (int32_t) address[0];
+    plasticity_trace_region_data.alpha = config->alpha;
 
     // Copy LUTs from following memory
-    address_t lut_address = &address[1];
+    address_t lut_address = config->lut_data;
```
Comment (rowleya) [inline neural_modelling/src/neuron/plasticity/stdp/timing_dependence/timing_vogels_2011_impl.c:49]: This is good; I wonder if it highlights a mistake below though!  I think this should return the address after reading the LUT.  Not changed here though so implies a mistake in master...

File: neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_one_term_impl.c [hunk 2]
```diff
@@ -36,27 +45,26 @@ address_t weight_initialise(
...
-                plasticity_weight_region_data[s].a2_plus,
-                plasticity_weight_region_data[s].a2_minus);
+                s, dtcm_copy[s].min_weight, dtcm_copy[s].max_weight,
+                dtcm_copy[s].a2_plus, dtcm_copy[s].a2_minus);
     }
     log_debug("weight_initialise: completed successfully");
 
     // Return end address of region
-    return (address_t) plasticity_word;
+    return (address_t) config;
```
Comment (rowleya) [inline neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_one_term_impl.c:69]: As with a previous comment, I think this highlights the error in the master code that this doesn't return where it has been read up to.  I suspect this is less of an issue here due to the fact that this is normally called last, but maybe it should do what it says!

File: neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_two_term_impl.c [hunk 3]
```diff
@@ -36,32 +47,32 @@ address_t weight_initialise(
...
-                plasticity_weight_region_data[s].a3_minus);
+                s, dtcm_copy[s].min_weight, dtcm_copy[s].max_weight,
+                dtcm_copy[s].a2_plus, dtcm_copy[s].a2_minus,
+                dtcm_copy[s].a3_plus, dtcm_copy[s].a3_minus);
     }
     log_debug("weight_initialise: completed successfully");
 
     // Return end address of region
-    return (address_t) plasticity_word;
+    return (address_t) config;
```
Comment (rowleya) [inline neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_two_term_impl.c:77]: Marking to again highlight a potential master issue.

File: neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_multiplicative_impl.c [hunk 4]
```diff
@@ -51,29 +60,26 @@ address_t weight_initialise(
...
-                weight_multiply_right_shift[s]);
+                s, dtcm_copy[s].min_weight, dtcm_copy[s].max_weight,
+                dtcm_copy[s].a2_plus, dtcm_copy[s].a2_minus, shift);
     }
 
     log_debug("weight_initialise: completed successfully");
 
     // Return end address of region
-    return (address_t) plasticity_word;
+    return (address_t) config;
```
Comment (rowleya) [inline neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_multiplicative_impl.c:84]: Potential master issue

File: neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_one_term_impl.c (hunk 2 above)
Comment (dkfellows) [reply neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_one_term_impl.c:69]: This returns the address immediately after where the last bit of configuration was read from. Note that we do `config++` in the loop, _including in the final step._

File: neural_modelling/src/neuron/plasticity/stdp/timing_dependence/timing_vogels_2011_impl.c (hunk 1 above)
Comment (dkfellows) [reply neural_modelling/src/neuron/plasticity/stdp/timing_dependence/timing_vogels_2011_impl.c:49]: The update to that is done by `maths_copy_int16_lut(&lut_address)` just below.

Comment (rowleya) [reply neural_modelling/src/neuron/plasticity/stdp/timing_dependence/timing_vogels_2011_impl.c:49]: Ah yes, phew!  I missed that it was passed by reference...

File: neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_one_term_impl.c (hunk 2 above)
Comment (rowleya) [reply neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_one_term_impl.c:69]: Ah good again, missed the ++ bit.

File: neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_two_term_impl.c (hunk 3 above)
Comment (rowleya) [reply neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_two_term_impl.c:77]: Ignore my misreading

Comment (dkfellows) [reply neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_additive_two_term_impl.c:77]: Both `master` and my code are correct. It's the `config` pointer that moves.

File: neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_multiplicative_impl.c (hunk 4 above)
Comment (rowleya) [reply neural_modelling/src/neuron/plasticity/stdp/weight_dependence/weight_multiplicative_impl.c:84]: Ignore, just me not reading it right.

**Review by rowleya — APPROVED** [review]
Ignore my silly comments.  This all looks good (and master isn't broken thankfully)

**Comment by andrewgait** [comment]
> Ignore my silly comments. This all looks good (and master isn't broken thankfully)

I'd like to think I'd have noticed STDP was broken in master before now... ;)

**Pull request merged**